test/subspace-test-service/src/mock.rs

```diff
@@ -517,21 +608,82 @@ fn log_new_block(block: &Block, used_time_ms: u128) {
     );
 }
 
+struct GossipWorkerPreliminary {
```

Comment:
This abstraction looks sort of peculiar. Instead of having this abstraction specific for the test, I feel like we can think about something that can be both in the production and test env, reducing the differences in between.

(This is to exemplify my point, it's incomplete)
```diff
diff --git a/crates/subspace-node/src/bin/subspace-node.rs b/crates/subspace-node/src/bin/subspace-node.rs
index abeada976..a0bf39a5b 100644
--- a/crates/subspace-node/src/bin/subspace-node.rs
+++ b/crates/subspace-node/src/bin/subspace-node.rs
@@ -17,7 +17,7 @@
 //! Subspace node implementation.

 use core_evm_runtime::AccountId as AccountId20;
-use cross_domain_message_gossip::{cdm_gossip_peers_set_config, GossipWorker};
+use cross_domain_message_gossip::{cdm_gossip_peers_set_config, GossipWorker, GossipWorkerBuilder};
 use domain_client_executor::ExecutorStreams;
 use domain_eth_service::provider::EthProvider;
 use domain_eth_service::DefaultEthConfig;
@@ -618,8 +618,7 @@ fn main() -> Result<(), Error> {
                             })
                     };

-                    let (gossip_msg_sink, gossip_msg_stream) =
-                        tracing_unbounded("cross_domain_gossip_messages", 100);
+                    let (mut gossip_worker_builder, gossip_msg_sink) = GossipWorkerBuilder::new();

                     let executor_streams = ExecutorStreams {
                         primary_block_import_throttling_buffer_size,
@@ -646,12 +645,11 @@ fn main() -> Result<(), Error> {
                         primary_chain_node.sync_service.clone(),
                         &primary_chain_node.select_chain,
                         executor_streams,
-                        gossip_msg_sink.clone(),
+                        gossip_msg_sink.clone()
                     )
                         .await?;

-                    let mut domain_tx_pool_sinks = BTreeMap::new();
-                    domain_tx_pool_sinks.insert(DomainId::SYSTEM, system_domain_node.tx_pool_sink);
+                    gossip_worker_builder.push_domain_tx_pool_sink(DomainId::SYSTEM, system_domain_node.tx_pool_sink);

                     primary_chain_node
                         .task_manager
@@ -698,7 +696,7 @@ fn main() -> Result<(), Error> {
                                         .clone(),
                                     select_chain: primary_chain_node.select_chain.clone(),
                                     executor_streams,
-                                    gossip_message_sink: gossip_msg_sink,
+                                    gossip_message_sink: gossip_msg_sink.clone(),
                                     provider: DefaultProvider,
                                 };

@@ -720,10 +718,8 @@ fn main() -> Result<(), Error> {
                                     >(core_domain_params)
                                         .await?;

-                                domain_tx_pool_sinks.insert(
-                                    core_domain_cli.domain_id,
-                                    core_domain_node.tx_pool_sink,
-                                );
+                                gossip_worker_builder.push_domain_tx_pool_sink(core_domain_cli.domain_id, core_domain_node.tx_pool_sink);
+
                                 primary_chain_node
                                     .task_manager
                                     .add_child(core_domain_node.task_manager);
@@ -752,7 +748,7 @@ fn main() -> Result<(), Error> {
                                         .clone(),
                                     select_chain: primary_chain_node.select_chain.clone(),
                                     executor_streams,
-                                    gossip_message_sink: gossip_msg_sink,
+                                    gossip_message_sink: gossip_msg_sink.clone(),
                                     provider: DefaultProvider,
                                 };

@@ -774,10 +770,7 @@ fn main() -> Result<(), Error> {
                                     >(core_domain_params)
                                         .await?;

-                                domain_tx_pool_sinks.insert(
-                                    core_domain_cli.domain_id,
-                                    core_domain_node.tx_pool_sink,
-                                );
+                                gossip_worker_builder.push_domain_tx_pool_sink(core_domain_cli.domain_id, core_domain_node.tx_pool_sink);
                                 primary_chain_node
                                     .task_manager
                                     .add_child(core_domain_node.task_manager);
@@ -824,7 +817,7 @@ fn main() -> Result<(), Error> {
                                     primary_network_sync_oracle: primary_chain_node.sync_service.clone(),
                                     select_chain: primary_chain_node.select_chain.clone(),
                                     executor_streams,
-                                    gossip_message_sink: gossip_msg_sink,
+                                    gossip_message_sink: gossip_msg_sink.clone(),
                                     provider: eth_provider,
                                 };

@@ -846,10 +839,7 @@ fn main() -> Result<(), Error> {
                                     >(core_domain_params)
                                         .await?;

-                                domain_tx_pool_sinks.insert(
-                                    core_domain_cli.domain_id,
-                                    core_domain_node.tx_pool_sink,
-                                );
+                                gossip_worker_builder.push_domain_tx_pool_sink(core_domain_cli.domain_id, core_domain_node.tx_pool_sink);
                                 primary_chain_node
                                     .task_manager
                                     .add_child(core_domain_node.task_manager);
@@ -864,10 +854,9 @@ fn main() -> Result<(), Error> {
                         }
                     }

-                    let cross_domain_message_gossip_worker = GossipWorker::<Block>::new(
+                    let cross_domain_message_gossip_worker = gossip_worker_builder.build::<Block, _, _>(
                         primary_chain_node.network_service.clone(),
                         primary_chain_node.sync_service.clone(),
-                        domain_tx_pool_sinks,
                     );

                     primary_chain_node
@@ -876,7 +865,7 @@ fn main() -> Result<(), Error> {
                         .spawn_essential_blocking(
                             "cross-domain-gossip-message-worker",
                             None,
-                            Box::pin(cross_domain_message_gossip_worker.run(gossip_msg_stream)),
+                            Box::pin(cross_domain_message_gossip_worker.run()),
                         );

                     system_domain_node.network_starter.start_network();
diff --git a/domains/client/cross-domain-message-gossip/src/gossip_worker.rs b/domains/client/cross-domain-message-gossip/src/gossip_worker.rs
index 4046974c2..c84d2aedd 100644
--- a/domains/client/cross-domain-message-gossip/src/gossip_worker.rs
+++ b/domains/client/cross-domain-message-gossip/src/gossip_worker.rs
@@ -7,7 +7,7 @@ use sc_network_gossip::{
     GossipEngine, MessageIntent, Syncing as GossipSyncing, ValidationResult, Validator,
     ValidatorContext,
 };
-use sc_utils::mpsc::TracingUnboundedSender;
+use sc_utils::mpsc::{tracing_unbounded, TracingUnboundedReceiver, TracingUnboundedSender};
 use sp_core::twox_256;
 use sp_domains::DomainId;
 use sp_runtime::traits::{Block as BlockT, Hash as HashT, Header as HeaderT};
@@ -28,11 +28,58 @@ pub struct Message {
     pub encoded_data: Vec<u8>,
 }

+pub struct GossipWorkerBuilder {
+    gossip_msg_stream: TracingUnboundedReceiver<Message>,
+    domain_tx_pool_sinks: BTreeMap<DomainId, DomainTxPoolSink>,
+}
+
+impl GossipWorkerBuilder {
+    pub fn new() -> (Self, TracingUnboundedSender<Message>) {
+        let (gossip_msg_sink, gossip_msg_stream) =
+            tracing_unbounded("cross_domain_gossip_messages", 100);
+        (
+            Self {
+                gossip_msg_stream,
+                domain_tx_pool_sinks: BTreeMap::new(),
+            },
+            gossip_msg_sink,
+        )
+    }
+
+    /// Collect the domain tx pool sink that will be used by the gossip message worker later.
+    pub fn push_domain_tx_pool_sink(
+        &mut self,
+        domain_id: DomainId,
+        tx_pool_sink: DomainTxPoolSink,
+    ) {
+        self.domain_tx_pool_sinks.insert(domain_id, tx_pool_sink);
+    }
+
+    pub fn build<Block, Network, GossipSync>(
+        self,
+        network: Network,
+        sync: Arc<GossipSync>,
+    ) -> GossipWorker<Block>
+    where
+        Block: BlockT,
+        Network: sc_network_gossip::Network<Block> + Send + Sync + Clone + 'static,
+        GossipSync: GossipSyncing<Block> + 'static,
+    {
+        let Self {
+            gossip_msg_stream,
+            domain_tx_pool_sinks,
+        } = self;
+
+        GossipWorker::new(network, sync, gossip_msg_stream, domain_tx_pool_sinks)
+    }
+}
+
 /// Gossip worker to gossip incoming and outgoing messages to other peers.
 /// Also, streams the decoded extrinsics to destination domain tx pool if available.
 pub struct GossipWorker<Block: BlockT> {
     gossip_engine: Arc<Mutex<GossipEngine<Block>>>,
     gossip_validator: Arc<GossipValidator>,
+    gossip_msg_stream: TracingUnboundedReceiver<Message>,
     domain_tx_pool_sinks: BTreeMap<DomainId, DomainTxPoolSink>,
 }

@@ -52,6 +99,7 @@ impl<Block: BlockT> GossipWorker<Block> {
     pub fn new<Network, GossipSync>(
         network: Network,
         sync: Arc<GossipSync>,
+        gossip_msg_stream: TracingUnboundedReceiver<Message>,
         domain_tx_pool_sinks: BTreeMap<DomainId, DomainTxPoolSink>,
     ) -> Self
     where
@@ -69,15 +117,13 @@ impl<Block: BlockT> GossipWorker<Block> {
         GossipWorker {
             gossip_engine,
             gossip_validator,
+            gossip_msg_stream,
             domain_tx_pool_sinks,
         }
     }

     /// Starts the Gossip message worker.
-    pub async fn run<GossipMessageStream>(mut self, mut gossip_msg_stream: GossipMessageStream)
-    where
-        GossipMessageStream: Stream<Item = Message> + Unpin + futures::stream::FusedStream,
-    {
+    pub async fn run(mut self) {
         let mut incoming_cross_domain_messages = Box::pin(
             self.gossip_engine
                 .lock()
@@ -99,7 +145,7 @@ impl<Block: BlockT> GossipWorker<Block> {
                     }
                 },

-                cross_domain_message = gossip_msg_stream.next().fuse() => {
+                cross_domain_message = self.gossip_msg_stream.next().fuse() => {
                     if let Some(msg) = cross_domain_message {
                         tracing::debug!(target: LOG_TARGET, "Incoming cross domain message for domain: {:?}", msg.domain_id);
                         self.handle_cross_domain_message(msg);
diff --git a/domains/client/cross-domain-message-gossip/src/lib.rs b/domains/client/cross-domain-message-gossip/src/lib.rs
index a8c03e605..12591b8b6 100644
--- a/domains/client/cross-domain-message-gossip/src/lib.rs
+++ b/domains/client/cross-domain-message-gossip/src/lib.rs
@@ -3,5 +3,7 @@
 mod gossip_worker;
 mod message_listener;

-pub use gossip_worker::{cdm_gossip_peers_set_config, DomainTxPoolSink, GossipWorker, Message};
+pub use gossip_worker::{
+    cdm_gossip_peers_set_config, DomainTxPoolSink, GossipWorker, GossipWorkerBuilder, Message,
+};
 pub use message_listener::start_domain_message_listener;
diff --git a/domains/test/service/src/system_domain.rs b/domains/test/service/src/system_domain.rs
index 5a21c0113..2295ad6e9 100644
--- a/domains/test/service/src/system_domain.rs
+++ b/domains/test/service/src/system_domain.rs
@@ -1,7 +1,7 @@
 //! Utilities used for testing with the system domain.
 #![warn(missing_docs)]
 use crate::{construct_extrinsic_generic, node_config};
-use cross_domain_message_gossip::GossipWorker;
+use cross_domain_message_gossip::{GossipWorker, GossipWorkerBuilder};
 use domain_client_executor::ExecutorStreams;
 use domain_service::{DomainConfiguration, FullPool};
 use frame_system_rpc_runtime_api::AccountNonceApi;
@@ -139,20 +139,17 @@ async fn run_executor_with_mock_primary_node(
         tx_pool_sink,
     } = system_domain_node;

-    let mut domain_tx_pool_sinks = BTreeMap::new();
-    domain_tx_pool_sinks.insert(DomainId::SYSTEM, tx_pool_sink);
-    let cross_domain_message_gossip_worker = GossipWorker::<Block>::new(
-        network_service.clone(),
-        sync_service.clone(),
-        domain_tx_pool_sinks,
-    );
+    let (mut gossip_worker_builder, _gossip_msg_sink) = GossipWorkerBuilder::new();
+    gossip_worker_builder.push_domain_tx_pool_sink(DomainId::SYSTEM, tx_pool_sink);
+    let cross_domain_message_gossip_worker =
+        gossip_worker_builder.build::<Block, _, _>(network_service.clone(), sync_service.clone());

     task_manager
         .spawn_essential_handle()
         .spawn_essential_blocking(
             "cross-domain-gossip-message-worker",
             None,
-            Box::pin(cross_domain_message_gossip_worker.run(gossip_msg_stream)),
+            Box::pin(cross_domain_message_gossip_worker.run()),
         );

     network_starter.start_network();
```

And then we have the `gossp_message_sink` and the builder, you can use the builder to actually build the worker when needed.

Reply:
Make sense to me, thanks!